baseapp/baseapp.go

```diff
@@ -300,6 +310,78 @@ func (app *BaseApp) QueryRouter() sdk.QueryRouter { return app.queryRouter }
 // GRPCQueryRouter returns the GRPCQueryRouter of a BaseApp.
 func (app *BaseApp) GRPCQueryRouter() *GRPCQueryRouter { return app.grpcQueryRouter }
 
+// RegisterGRPC registers gRPC services directly with the gRPC server.
+func (app *BaseApp) RegisterGRPC(server gogogrpc.Server) {
+	// Define an interceptor for all gRPC queries: this interceptor will create
+	// a new sdk.Context, and pass it into the query handler.
+	interceptor := func(grpcCtx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+		// To create the correct sdk.Context, we first need to transform the
+		// req into an abci.RequestQuery
+		msg, ok := req.(proto.Message)
+		if !ok {
+			return nil, status.Errorf(codes.Internal, "unable to proto marshal")
+		}
+
+		msgBz, err := proto.Marshal(msg)
+		if err != nil {
+			return nil, err
+		}
+
+		abciReq := abci.RequestQuery{
+			Data: msgBz,
+			Path: info.FullMethod,
+		}
+
+		// Create the sdk.Context
+		sdkCtx, err := app.createQueryContext(abciReq)
+		if err != nil {
+			return nil, err
+		}
+
+		// Wrap the sdk.Context into a new context.Context
+		newCtx := sdk.WrapSDKContext(sdkCtx)
```

Comment:
Instead of using `WrapSDKContext` let's just have a function that sets `sdk.Context` on the context that came from gRPC. Then you don't need to worry about setting streams on that the context.